### PR TITLE
[Torch, QNN] Add support for quantized models via QNN

### PR DESCRIPTION
**Pull request by masahi**
This adds support for converting quantized models from torch to Relay via QNN. I believe this makes a good case for introducing Torch frontend to TVM, since the quantized Torch models -> ONNX path is not supported at the moment (or any time soon). Also torch cannot run quantized models on GPU yet, but we can run quantized models on CUDA very fast.

So far I've tested on quantized models from torchvision available at https://github.com/pytorch/vision/tree/master/torchvision/models/quantization (shufflenet is removed since it has issues with tracing), and a custom mobilenet v3 model.

Per channel quantization is fully supported, and it is a big win on mobilenet v3 (more than 10 point accuracy on imagenet). I think even TFLite or MXNet + QNN doesn't support per channel quantization end to end.

The PR contains 
* The converter to QNN
* Tests for individual quantized modules
* Tests on pretrained imagenet models with dummy calibration

Verifying the result is tricky due to difference in numerics between torch and tvm. 

Quantized modules tests are run but there is no assertion on how close the raw floating point output is to torch. I dumped some statistics that we can eye ball and verify that the result is reasonable.

 For imagenet tests, I use the real image and verify that two top3 label sets are identical. Here, I'm ignoring the order among top3, but for the 4 models that I added in the test, the order is actually correct. The order is ignored to make CI less fragile.

Below is the result on 1k subset of imagenet val data with full calibration + per channel quantization. I'm preparing a proper repo to do the same evaluation.

(I've just realized that digits after 0 seem weird... only 00, 33 and 67. Looking into it. I'm using the same evaluation code as Torch quantization tutorial https://pytorch.org/tutorials/advanced/static_quantization_tutorial.html, it seems they have the same issue)

model name | Torch-Top1 | Torch-Top5 | TVM-Top1 | TVM-Top5
-- | -- | -- | -- | --
resnet18 | 77.33 | 95.00 | 76.00 | 94.67
resnet50 | 82.67 | 97.00 | 81.67 | 96.67
inception_v3 | 78.33 | 94.67 | 78.67 | 94.67
googlenet| 78.33 | 95.33 | 78.33 | 95.67
mobilenet_v2 | 76.00 | 94.33 | 77.00 | 94.33
mobilenet_v3 | 72.00| 90.67 | 70.00 | 91.33

cc @anijain2305 @vinx13 @FrozenGene @jwfromm @alexwong @ajtulloch @hlu1 @yinghai

### CODE REVIEW

**Comment by masahi** [comment]
Test passed!!

File: python/tvm/relay/frontend/pytorch.py [hunk 1]
```diff
@@ -146,6 +149,10 @@ def _impl(inputs, input_types):
 def _relu():
     def _impl(inputs, input_types):
         data = inputs[0]
+        if input_types[0] == "quint8":
+            assert len(inputs) == 3, "Input quant param not found in op inputs"
+            input_zero_point = _expr.const(inputs[2])
```
Comment (anijain2305) [inline python/tvm/relay/frontend/pytorch.py]: Do you want to add a datatype "int32" here?

File: python/tvm/relay/frontend/qnn_torch.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+import numpy as np
+
+import tvm
+from tvm import relay
+from tvm.relay import expr as _expr
+from tvm.relay import op as _op
+from tvm.relay.frontend.common import infer_shape
+
+
+class QuantParam:
```
Comment (anijain2305) [inline python/tvm/relay/frontend/qnn_torch.py]: For consistency, I would suggest using QNN instead of quant wherever possible. We tried to do the same in TFLite and MXNet parser.

**Comment by masahi** [comment]
@anijain2305 Where do you want to see more comments? Happy to add them.

 I think the most non-obvious aspect of the code is the need to do some surgery on an input graph, to make all input and output quant params visible on the op inputs. Is that part clear?

File: tests/python/frontend/pytorch/qnn_test.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+
+        print("Torch top3 label:", pt_top3_labels)
+        print("TVM top3 label:", tvm_top3_labels)
+
+    for (model_name, pt_result, tvm_result) in results:
+        max_abs_diff = np.max(np.abs(tvm_result - pt_result))
+        mean_abs_diff = np.mean(np.abs(tvm_result - pt_result))
+        num_identical = np.sum(tvm_result == pt_result)
+
+        print("\nModel name: %s" % model_name)
```
[the remark this answers is not in the document]
Comment (masahi) [reply tests/python/frontend/pytorch/qnn_test.py]: Yeah, this is my pain point :) I added a sample output in https://github.com/apache/incubator-tvm/pull/4977/commits/a6239e19b430b1f069010e94dac939743dd36ff5

The problem is how close the result is to torch differs widely among different network and which of per tensor/per channel quantization is used. I could add something like `max_abs_diff` < 2.5, but I don't think that is too helpful.

I believe there is a room for improvement in terms of making the result closer to torch. Also quantization in general is WIP on torch as well. There are some quantized ops that piggy backs to fp32 by dequantize -> fp32 op -> quantize, which defeats the purpose of doing quantization (going faster than fp32).

I want to say this is an initial support for torch quantization by someone who is new to the quantization space, and that by making it public I am hoping that people can improve on it :)

**Comment by masahi** [comment]
@anijain2305 Added more comments, I think it is enough for now.

**Comment by anijain2305** [comment]
Will wait for a day or two for others to review.

**Comment by masahi** [comment]
@anijain2305 I think it is a good time to land it.

**Comment by anijain2305** [comment]
Thanks @masahi @jwfromm This is merged!

**Comment by masahi** [comment]
Hmm for some reason, this commit became @anijain2305' commit. 
https://github.com/apache/incubator-tvm/commits?author=anijain2305

Also see the commit log. Something has changed since today. Have we changed some settings or is this github bug? @tqchen

**Comment by greysteil** [comment]
FYI, this issue (the change in commit author) got escalated to me at GitHub. We have a bug in our squash and merge logic right now (introduced yesterday) which causes the original PR author to be removed from the list of commit co-authors in some cases. We're working on a fix now.

**Comment by tqchen** [comment]
This PR is among one of the PRs affected by the github squash commit bug. We take every contribution serious in the TVM community.  The community has decided to use revert/redo approach to amend the contributions as per https://github.com/apache/incubator-tvm/issues/5015
